README.md
# Site Endpoints Plugin

Copy the siteEndpoints.yaml into your `config/plugins`. Generate your recaptcha v3 key on the recaptcha site, replace `your-key-here` with your secret

The **Site Endpoints** Plugin is an extension for [Grav CMS](http://github.com/getgrav/grav). Composable Site Endpoints plugin

## Installation

Installing the Site Endpoints plugin can be done in one of three ways: The GPM (Grav Package Manager) installation method lets you quickly install the plugin with a simple terminal command, the manual method lets you do so via a zip file, and the admin method lets you do so via the Admin Plugin.

### GPM Installation (Preferred)

To install the plugin via the [GPM](http://learn.getgrav.org/advanced/grav-gpm), through your system's terminal (also called the command line), navigate to the root of your Grav-installation, and enter:

    bin/gpm install siteEndpoints

This will install the Site Endpoints plugin into your `/user/plugins`-directory within Grav. Its files can be found under `/your/site/grav/user/plugins/siteEndpoints`.

### Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `siteEndpoints`. You can find these files on [GitHub](https://github.com//grav-plugin-siteEndpoints) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/siteEndpoints

> NOTE: This plugin is a modular component for Grav which may require other plugins to operate, please see its [blueprints.yaml-file on GitHub](https://github.com//grav-plugin-siteEndpoints/blob/master/blueprints.yaml).

### Admin Plugin

If you use the Admin Plugin, you can install the plugin directly by browsing the `Plugins`-menu and clicking on the `Add` button.

## Configuration

Before configuring this plugin, you should copy the `user/plugins/siteEndpoints/siteEndpoints.yaml` to `user/config/plugins/siteEndpoints.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
```

Note that if you use the Admin Plugin, a file with your configuration named siteEndpoints.yaml will be saved in the `user/config/plugins/`-folder once the configuration is saved in the Admin.

## Usage

**Describe how to use the plugin.**

## Credits

**Did you incorporate third-party code? Want to thank somebody?**

## To Do

- [ ] Future plans, if any


CHANGELOG.md
# v0.0.2

## 06/24/2020

1. [](#new)
   - Update License and changelog

# v0.0.1

## 06/20/2020

1. [](#new)
   - Initial release


endpoints.php
<?php
namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Grav\Common\Grav;
use Grav\Common\Plugin;

class EndpointsPlugin extends Plugin
{
    protected $route = 'api';

    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onPluginsInitialized' => [
                ['autoload', 100000],
            ],
            'onPagesInitialized' => ['onPagesInitialized', 0],
        ];
    }

    /**
     * [onPluginsInitialized:100000] Composer autoload.
     *
     * @return ClassLoader
     */
    public function autoload()
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    public function onPagesInitialized()
    {
        $uri = $this->grav['uri'];
        $config = $this->config->get('plugins.endpoints');

        if (strpos($uri->path(), $this->route) === false) {
            return;
        }

        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];

        if ($resource) {
            $file = __DIR__ . '/resources/' . $resource . '.php';
            if (file_exists($file)) {
                require_once $file;
                $resourceClassName = '\Grav\Plugin\Endpoints\\' . ucfirst($resource);
                $resource = new $resourceClassName($this->grav);
                $output = $resource->execute($config);
                $resource->setHeaders();

                echo $output;
            } else {
                header('HTTP/1.1 404 Not Found');
            }
        }

        exit();
    }
}


siteEndpoints.php
<?php
namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Grav\Common\Grav;
use Grav\Common\Plugin;

class SiteEndpointsPlugin extends Plugin
{
    protected $route = 'api';

    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onPluginsInitialized' => [
                ['autoload', 100000],
            ],
            'onPagesInitialized' => ['onPagesInitialized', 0],
        ];
    }

    /**
     * [onPluginsInitialized:100000] Composer autoload.
     *
     * @return ClassLoader
     */
    public function autoload()
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    public function onPagesInitialized()
    {
        $uri = $this->grav['uri'];
        $config = $this->config->get('plugins.siteEndpoints');

        if (strpos($uri->path(), $this->route) === false) {
            return;
        }

        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];

        if ($resource) {
            $file = __DIR__ . '/resources/' . $resource . '.php';
            if (file_exists($file)) {
                require_once $file;
                $resourceClassName = '\Grav\Plugin\SiteEndpoints\\' . ucfirst($resource);
                $resource = new $resourceClassName($this->grav);
                $output = $resource->execute($config);
                $resource->setHeaders();

                echo $output;
            } else {
                header('HTTP/1.1 404 Not Found');
            }
        }

        exit();
    }
}


siteendpoints.php
<?php
namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Grav\Common\Grav;
use Grav\Common\Plugin;

class SiteEndpointsPlugin extends Plugin
{
    protected $route = 'api';

    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onPluginsInitialized' => [
                ['autoload', 100000],
            ],
            'onPagesInitialized' => ['onPagesInitialized', 0],
        ];
    }

    /**
     * [onPluginsInitialized:100000] Composer autoload.
     *
     * @return ClassLoader
     */
    public function autoload()
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    public function onPagesInitialized()
    {
        $uri = $this->grav['uri'];
        $config = $this->config->get('plugins.siteendpoints');

        if (strpos($uri->path(), $this->route) === false) {
            return;
        }

        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];

        if ($resource) {
            $file = __DIR__ . '/resources/' . $resource . '.php';
            if (file_exists($file)) {
                require_once $file;
                $resourceClassName = '\Grav\Plugin\SiteEndpoints\\' . ucfirst($resource);
                $resource = new $resourceClassName($this->grav);
                $output = $resource->execute($config);
                $resource->setHeaders();

                echo $output;
            } else {
                header('HTTP/1.1 404 Not Found');
            }
        }

        exit();
    }
}


resources/captcha.php
<?php
namespace Grav\Plugin\SiteEndpoints;

use Grav\Common\Grav;
use Grav\Common\Plugin;
use Grav\Common\Uri;
use ReCaptcha\ReCaptcha;
use ReCaptcha\RequestMethod\CurlPost;

require_once 'resource.php';

/**
 * Captcha Verification API
 */
class Captcha extends Resource
{

    /**
     * Post Captcha
     *
     * Implements:
     *
     * - POST /api/captcha
     * - @return response
     */
    public function postItem($config)
    {

        $secret = $config['recaptcha']['secret_key'];

        $data = $this->getPost();
        $uri = $this->grav['uri'];
        $action = $data->action;
        $hostname = $uri->host();
        $ip = Uri::ip();

        $recaptcha = new ReCaptcha($secret);

        if (extension_loaded('curl')) {
            $recaptcha = new ReCaptcha($secret, new CurlPost());
        }

        // Add version 3 specific options

        $token = $data->token;
        $resp = $recaptcha
            ->setExpectedHostname($hostname)
            ->setExpectedAction($action)
            ->setScoreThreshold(0.5)
            ->verify($token, $ip);

        if (!$resp->isSuccess()) {
            $errors = $resp->getErrorCodes();

            $this->grav['log']->addWarning('Form reCAPTCHA Errors: [' . $uri->route() . '] ' . json_encode($errors));

            return json_encode($errors);
        }

        return json_encode($resp->isSuccess());

    }
}


resources/version.php
<?php
namespace Grav\Plugin\SiteEndpoints;

require_once 'resource.php';

use RocketTheme\Toolbox\File\File;
use Symfony\Component\Yaml\Yaml;

/**
 * Version API
 */
class Version extends Resource
{
    /**
     * Get API Version
     *
     * Implements:
     *
     * - GET /api/version
     */
    public function getList()
    {
        $version = $this->getVersion();
        return strtok($version, '.');
    }

    /**
     * Implements:
     *
     * - GET /api/version/major
     * - GET /api/version/minor
     * - GET /api/version/full
     */
    public function getItem()
    {
        switch ($this->getIdentifier()) {
            case 'major':
                // Get API Major Version
                $version = $this->getVersion();
                return strtok($version, '.');
            case 'minor':
                // Get API Minor Version
                $version = $this->getVersion();
                return strtok($version, '.') . '.' . strtok('.');
            case 'full':
                //Get API Full Version
                return $this->getVersion();
        }

        return;
    }

    /**
     * Fetch the plugin version from the blueprints.yaml file
     */
    private function getVersion()
    {
        $locator = $this->grav['locator'];
        $path = $locator->findResource('user://plugins/endpoints', true);
        $fullFileName = $path . DS . 'blueprints.yaml';

        $file = File::instance($fullFileName);
        $data = Yaml::parse($file->content());

        return $data['version'];
    }
}


resources/resource.php
<?php
namespace Grav\Plugin\SiteEndpoints;

use Grav\Common\Grav;

class Resource
{
    /**
     * Constructor.
     */
    public function __construct(Grav $grav)
    {
        $this->grav = $grav;
    }

    /**
     * Execute the request
     *
     * @return string the output to render
     */
    public function execute($config = null)
    {
        $httpMethod = $this->getMethod();

        if ($httpMethod == 'get') {
            if ($this->getIdentifier($config)) {
                $return = $this->getItem();
            } else {
                $return = $this->getList($config);
            }
        } else {
            $method = $httpMethod . 'Item';
            $return = $this->$method($config);
        }

        if ($return !== '') {
            return json_encode($return);
        } else {
            return '';
        }

    }

    /**
     * Get the request method
     *
     * @return string the method name, lowercase
     */
    protected function getMethod()
    {
        return strtolower($_SERVER['REQUEST_METHOD']);
    }

    /**
     * Get the resource name
     *
     * @return string the resource name
     */
    protected function getResource()
    {
        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];
        return $resource;
    }

    /**
     * Get the identifier name
     *
     * @return string the resource identifier name
     */
    protected function getIdentifier()
    {
        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 2);

        $identifier = join('/', $paths);
        return $identifier;
    }

    /**
     * Set the response headers based on the content retrieved
     *
     * Use the following response headers:
     * - ETag: An arbitrary string for the version of a representation. Make sure to include the media type in the hash value, because that makes a different representation. (ex: ETag: "686897696a7c876b7e")
     * - Date: Date and time the response was returned (in RFC1123 format). (ex: Date: Sun, 06 Nov 1994 08:49:37 GMT)
     * - Cache-Control: The maximum number of seconds (max age) a response can be cached. However, if caching is not supported for the response, then no-cache is the value. (ex: Cache-Control: 360 or Cache-Control: no-cache)
     * - Expires: If max age is given, contains the timestamp (in RFC1123 format) for when the response expires, which is the value of Date (e.g. now) plus max age. If caching is not supported for the response, this header is not present. (ex: Expires: Sun, 06 Nov 1994 08:49:37 GMT)
     * - Last-Modified: The timestamp that the resource itself was modified last (in RFC1123 format). (ex: Last-Modified: Sun, 06 Nov 1994 08:49:37 GMT)
     *
     * @todo implement
     */
    public function setHeaders()
    {
        header('Content-type: application/json');

        // Calculate Expires Headers if set to > 0
        $expires = $this->grav['config']->get('system.pages.expires');
        if ($expires > 0) {
            $expires_date = gmdate('D, d M Y H:i:s', time() + $expires) . ' GMT';
            header('Cache-Control: max-age=' . $expires);
            header('Expires: ' . $expires_date);
        }

        // TODO: Set Last-Modified
        // TODO: Set the ETag
        // TODO: Set the HTTP response code
        // TODO: Set the Date
    }

    /**
     * Set the correct error code
     *
     * @todo implement all error codes
     */
    protected function setErrorCode($code)
    {
        switch ((int) $code) {
            case 401:
                header('HTTP/1.1 401 Unauthorized');
                break;
            case 403:
                header('HTTP/1.1 403 Forbidden');
                break;
            case 404:
                header('HTTP/1.1 404 Not Found');
                break;
            case 405:
                header('HTTP/1.1 405 Not Allowed');
                break;
            case 501:
                header('HTTP/1.1 501 Not Implemented');
                break;
            default:
                header('HTTP/1.1 ' . $code);

        }
    }

    /**
     * Get list action
     *
     * @return
     */
    public function getList($config)
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Get item action
     *
     * @return
     */
    public function getItem($config)
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Post action
     *
     * @return
     */
    public function postItem($config)
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Put action
     *
     * @return
     */
    public function putItem($config)
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Delete action
     *
     * @return
     */
    public function deleteItem($config)
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Prepare and return POST data.
     *
     * @param array $post
     * @return array
     */
    protected function getPost()
    {
        return json_decode(file_get_contents('php://input'));
    }

    /**
     * Builds the return message.
     *
     * Example usage from a resource action:
     *
     *   ```
     *   $this->setErrorCode(403);
     *   $message = $this->buildReturnMessage('Page already exists. Cannot create a page with the same route');
     *   return $message;
     *   ```
     *
     * @param array $post
     * @return array
     */
    protected function buildReturnMessage($message)
    {
        return [
            'message' => $message,
        ];
    }

}
